tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+
+        # select nearest cell based on euclidian distance
+        nearest = np.unravel_index(np.argmin(np.sqrt((
+                                   coords[0] - x)**2 + (coords[1] - y)**2)),
+                                   self.shape)
+        return nearest[1], nearest[0]
+
+    def flowdir(self, data=None, include_edges=True, dirmap=(1,2,3,4,5,6,7,8),
+                nodata=0, flat=-1, inplace=True):
+        """
+        Generates a flow direction grid from a DEM grid.
+
+        Parameters
+        ----------
+        data : numpy ndarray
+               Array representing DEM grid
+        include_edges : bool
+                        Whether to include outer rim of grid.
+        dirmap : list or tuple
+                 List of integer values representing the following
+                 cardinal and intercardinal directions (in order):
+                 [N, NE, E, SE, S, SW, W, NW]
+        """
+
+        # if data not provided, use self.dem
+        if data is None:
+            if hasattr(self, 'dem'):
+                data = self.dem
+
+        # generate grid of indices
+        indices = np.indices(self.shape, dtype=self.shape_min)
+
+        # handle nodata values in dem
+        dem_nodata = self.nodata['dem']
+        dem_mask = (data == dem_nodata)        
+        np.place(data, dem_mask, np.iinfo(data.dtype.type).max)
+
+        # initialize indices of corners
+        corner = {
+        'nw' : {'k' : tuple(indices[:,0,0]),
+                'v' : [[0,1,1], [1,1,0]],
+                'pad': np.array([3,4,5])},
+        'ne' : {'k' : tuple(indices[:,0,-1]),
+                'v' : [[1,1,0], [-1,-2,-2]],
+                'pad': np.array([5,6,7])},
+        'sw' : {'k' : tuple(indices[:,-1,0]),
+                'v' : [[-2,-2,-1], [0,1,1]],
+                'pad': np.array([1,2,3])},
+        'se' : {'k' : tuple(indices[:,-1,-1]),
+                'v' : [[-1,-2,-2], [-2,-2,-1]],
+                'pad': np.array([7,8,1])}
+        }
+    
+        # initialize indices of edges
+        edge = {
+        'n' : {'k' : tuple(indices[:,0,1:-1]),
+               'pad' : np.array([3,4,5,6,7])},
+        'w' : {'k' : tuple(indices[:,1:-1,0]),
+               'pad' : np.array([1,2,3,4,5])},
+        'e' : {'k' : tuple(indices[:,1:-1,-1]),
+               'pad' : np.array([1,5,6,7,8])},
+        's' : {'k' : tuple(indices[:,-1,1:-1]),
+               'pad' : np.array([1,2,3,7,8])}
+        }
+    
+        # initialize indices of body (all cells except edges and corners)
+        body = indices[:, 1:-1, 1:-1]
+    
+        # initialize output array
+        outmap = np.full(self.shape, nodata, dtype=np.int8)
+    
+        # select the eight cells surrounding a cell
+        def select_surround(i, j):
+            return ([i-1, i-1, i+0, i+1, i+1, i+1, i+0, i-1],
+                   [j+0, j+1, j+1, j+1, j+0, j-1, j-1, j-1])
+    
+        # select the five cells surrounding an edge cell
+        def select_edge_sur(k):
```

Comment:
consider moving these `select_` functions out of this method to allow for easier testing.